### DISCUSSION TITLE
[API discussion] inViewport prop

### DISCUSSION THREAD

**Discussion by mattgperry** [motiondivision/motionone#2] [Ideas]
Hey all. Today I implemented an `inViewport` prop for the React API. It's not published yet - before I do I thought I'd get people's thoughts. 

Currently, it looks like this:

```jsx
<animated.div
  style={{ opacity: 0 }}
  inViewport={{ opacity: 1 }}
  // Optional props
  viewport={parentRef}
  viewportThreshold={0.5}
  viewportMargin="100px"
  onViewportEnter={() => {}}
  onViewportLeave={() => {}}
/>
```

What do people think about `inViewport`? I think it works for trigger, binary states but I wonder if it'll be confusing when potentially looking at `ScrollTimeline`-based APIs.

Additionally we have these namespaced props. How about a straight pass-through of `IntersectionObserverInit` props:

```jsx
<animated.div viewport={{ root, rootMargin, threshold }} />
```

I'm also thinking about how we'd support multiple thresholds. Perhaps an array syntax:

```jsx
<animated.div
  inViewport={[
    { scale: 1 },
    { rotate: 90 }
  ]}
  viewport={{ thresholds: [0, 0.5] }}
/>
```

Your comments are appreciated!

**Comment by okikio** [comment]
You might want to use `ScrollTimeline` instead of `IntersectionObserver`. `ScrollTimeline`'s use cases are very broad and powerful, plus there is already a polyfill out, if you want to support older browsers.

[https://www.bram.us/2021/03/04/the-future-of-css-scroll-linked-animations-part-2/#scroll-timeline-and-element-based-offsets](https://www.bram.us/2021/03/04/the-future-of-css-scroll-linked-animations-part-2/#scroll-timeline-and-element-based-offsets)

**Comment by mattgperry** [comment]
I think because the `ScrollTimeline` spec is still in flux and this specific use-case is animation triggering (rather than driving) `IntersectionObserver` is better for this, but I do intend on providing a `ScrollTimeline` wrapper at some point for sure.

**Reply by loganjamesnottbohm** [reply]
Is there currently a way to incorporate `ScrollTimeline` with Motion animations or will this only be possible once the proposed wrapper is implemented?

**Reply by mattgperry** [reply]
Yeah it’s not possible for now. I’m keeping clear of that for a while, at least until there’s a stable spec proposal. It could be that I take a look at a custom solution that could one day be moved to ScrollTimeline but it wouldn’t actually be using ScrollTimeline in the meantime

**Comment by lunelson** [comment]
A couple of thoughts:

1. the _direction_ of the scroll would be a nice addition—perhaps as a parameter passed to the `onViewportEnter` and `onViewportLeave` handlers (might be a `1` or `-1`), or as separate handlers like `onViewportEnterBack` and `onViewPortLeaveBack` (GSAP does something like this)
2. if you're considering a `ScrollTimeine`-driven API for _scroll-driven_ animation in the future, you might consider aligning on the threshold terminology now for _scroll-triggers_, as they come up in `ScrollTimeline` anyway and they reference current directional CSS terminology, i.e. `start`/`end` rather than `top`/`bottom`—see [this section](https://www.bram.us/2021/03/04/the-future-of-css-scroll-linked-animations-part-2/#combining-offsets)

**Comment by subtleGradient** [comment]
Lately I've been playing around with VisualViewport to implement some zoom-oriented stuff. 
Seems like that could integrate with `inViewport` fantastically!
